src/Kestrel.Core/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
+{
+    public class Http2OutputProducer : IHttpOutputProducer
+    {
+        private static readonly ArraySegment<byte> _emptyData = new ArraySegment<byte>(new byte[0]);
+
+        private readonly int _streamId;
+        private readonly IHttp2FrameWriter _frameWriter;
+
+        public Http2OutputProducer(int streamId, IHttp2FrameWriter frameWriter)
+        {
+            _streamId = streamId;
+            _frameWriter = frameWriter;
+        }
+
+        public void Dispose()
+        {
+        }
+
+        public void Abort(Exception error)
+        {
+            // TODO: RST_STREAM?
+        }
+
+        public Task FlushAsync(CancellationToken cancellationToken) => _frameWriter.FlushAsync(cancellationToken);
+
+        public Task Write100ContinueAsync(CancellationToken cancellationToken) => _frameWriter.Write100ContinueAsync(_streamId);
+
+        public Task WriteDataAsync(ArraySegment<byte> data, bool chunk, CancellationToken cancellationToken)
+        {
+            if (chunk)
+            {
+                throw new ArgumentException("Chunked transfer coding is not supported in HTTP/2", nameof(chunk));
```

Comment:
There was a short time where we passed the chunk bool to WriteAsync in order to allow chunks to be aggregated, but now OutputProducer.WriteAsync just adds the chunking inline.

So instead of adding the chunk param to the shared IHttpOutputProducer interface, I would just add the chunks in Frame.WriteChunkedAsync.

Reply:
974e5532083fa546237b93459c0bdc0beb04d5c7

I prefer this to not having the `chunk` parameter in the interface, but it's still protocol-specific in that `Http2OutputProducer` doesn't implement the new `WriteAsync` method.